### ISSUE TITLE
Artifact Publishing disproportionately slow for large number of small files

### ISSUE THREAD

**Issue by gregpakes** [microsoft/azure-pipelines-agent#1480]
> I have published this over here too, I'll delete whichever one is not appropriate.
> https://github.com/Microsoft/vsts-tasks/issues/6830

I am using VSTS with the hosted agent.

Artifact publishing is very slow.  Cloning the repo, which is much larger than the artifacts that need publishing takes about 3-4 mins.  Publishing the artifacts is taking **23 minutes**.

Here are the details:

- 23k files in the final artifacts that need publishing
- I've seen that people suggest zipping it up, then publishing.  We have tried this before, but the zipping process takes longer than the publishing.  Then I also need to unzip them in the release process.
- I am publishing to the Server, not to a file share.
- I am using a hosted agent.

I've read on some forums that it takes a long time because it only does 4 files at a time.  

Is there any way to speed this up?

**Comment by gregpakes** [comment]
I've figured a decent work around for this.  The publish now takes 28 seconds.

I use the following task [Artifact Compressor](https://marketplace.visualstudio.com/items?itemName=roshkovski.2B9619D5-7BE9-4ED7-BF10-707CB17F657A) which allows you to create a zip file with **no compression**.  This obviously solves the problem described above where the compression takes too long.

Publishing the single zip archive is much faster.

This has taken 30 mins off my build time, halving it!

**Comment by bryanmacfarlane** [comment]
@madhurig - cpu bound on large number of small files?  I think it's more about the target machine than the task??

**Comment by gregpakes** [comment]
@bryanmacfarlane

Why do you think it’s CPU bound?  Also I have no control over the machines.  This is hosted build agent publishing to VSTS container

**Comment by bryanmacfarlane** [comment]
I read your statement as it's faster with no compression.  But I see now that you're saying the fact that you went to one file (without spending cycles on compressing) you're faster.

If that's correct, then very likely it's because it's not writing metadata to the backend for each file (we support browsing a drop in the web UI).  Just one metadata row. It's also taking the overhead of connecting and starting a stream for each file down to one.   That's especially true for a very large number of small files where the metadata writes and overhead per file dominates the time.

Upside is for one zip is (for large drops) it's faster.  Downside, you can't browse the drop and download portions of the drop as a zip (UI supports browsing to a sub folder and downloading).  

Side note that if you want to try your dataset on a VM where you do have access and control, the hosted agent is a DS2V2 or DS2V3 azure VM - nothing magical :)  We also do offer options where you do have control over the machines (deploy your own agent on VM you acquire): https://docs.microsoft.com/en-us/vsts/build-release/actions/agents/v2-windows?view=vsts

**Comment by bryanmacfarlane** [comment]
Yes, I understand.  I added @madhurig because there's investment going on in this area to rework artifacts.  Her team is driving it.  I was trying to outline where I think it's blocking :)

**Comment by bryanmacfarlane** [comment]
Can you confirm whether it's a large number of small files?

**Comment by gregpakes** [comment]
@bryanmacfarlane 

Its 23K files with a total size of 400MB.

**Comment by bryanmacfarlane** [comment]
Yeah, that's pretty much the worst case for that task that's bound by metadata writes.

**Comment by ChristopherHaws** [comment]
We also have this same issue with our ASP.NET site which we build and publish using VSTS. We tried compressing the files but it made it so that we couldn't select which artifacts to download in our release (our website and our automated UI tests are in the artifacts). We at one point setup multiple artifacts but it became to complicated to follow so we simply stopped using compression. It also made it so that we couldn't select files from the artifacts using the release file pickers.

It would be nice to have the option to compress artifacts in the `Publish Build Artifacts` task, and have VSTS simply know how to handle compressed artifacts. Even if the compression level is set to "Store" it makes a huge speed improvement. It's more about lowering the amount of files that are getting uploaded to VSTS than the size of the files.

**Comment by gregpakes** [comment]
@ChristopherHaws - I have this working very well by compressing artifacts separately and publishing them separately.   I can then select which artifacts to download (and decompress) in RM.  This works well and has reduced build/release time by 50%.

**Comment by Gold14388** [comment]
@gregpakes, I get an error saying that "[error]Microsoft.PowerShell.Commands.WriteErrorException: Failed to move D:\a\1\a\UDC.Framework.Core.deploy-readme.txt to D:\a\1\a\file.deploy-readme.txt.zip"
Any suggestions?

**Comment by gregpakes** [comment]
@Gold14388 
I assume you are using the Artifacts Compressor task.  I did not write this task and sadly the source is not on github.  I have asked the author to put the source on GitHub, but he doesn't want to.  If it was, I would be able to diagnose the issue easily. 

I have asked him about it here:  https://marketplace.visualstudio.com/items?itemName=roshkovski.2B9619D5-7BE9-4ED7-BF10-707CB17F657A#qna

I recommend you ask him a Q&A question on the link above and then also ask him to put the source on github.

**Comment by odministrator** [comment]
@Gold14388 could you please try again and share results?

**Comment by gregpakes** [comment]
We suddenly have the same issue on our builds....

**Comment by Gold14388** [comment]
@gregpakes Thanks for acknowledging my question. You too get the same error now? That's interesting. I wonder if it matters to have $(Build.ArtifactStagingDirectory) as source and destination for the task

**Comment by odministrator** [comment]
I guess that the problem can occur if you have file path instead of directory paths in source\destination. The idea of the task is to pack everything in zip files in build "a" directory and revert changes later during deployment. If you need something to archive file - there are Archive\Extract tasks by Microsoft which provide even better functionality and support tar\rar\7z also.
In my case everything works pretty fine and archives directories\files on build and expand on release side.

**Comment by gregpakes** [comment]
@oadministrator

I understand what you’re saying, but you have changed the functionality and broken a number of people. Could this have been released as a v2 and still allow is to use v1?

So are you saying that you’re not going to fix it?

Also the Microsoft tasks don’t work for me as they don’t support zero compression (as far as I know), which is my use case.

**Comment by odministrator** [comment]
Looks like the easiest way to manage the issue is create new version, as I haven't seen build logs and can't reproduce it.
I moved new script and task to v2 version and rolled-back changes to previous version in v1. So, old builds should be good now.
Please check v2 and in case of issues send me build logs in debug mode or at least describe scenario to reproduce.

**Comment by gregpakes** [comment]
@odministrator 

Much appreciated.

**Comment by gregpakes** [comment]
@odministrator 

This still isn't working for me.  The task version is:

Version      : 1.3.2

**Comment by odministrator** [comment]
Sorry, my fault. I commited minor version lower than it had been published yesterday.
Should be fixed in a minute (publish build is running).

**Comment by Gold14388** [comment]
@odministrator, @gregpakes, I used this link https://marketplace.visualstudio.com/items?itemName=roshkovski.2B9619D5-7BE9-4ED7-BF10-707CB17F657A#overview to get the latest version and I can only see version 1.2.1. Where can i find the latest version?

**Comment by odministrator** [comment]
@Gold14388 You are looking in a wrong place :) You check extension version, but we are talking about compress task version which is actually 1.5.0 for v1 and 2.0.0 for v2.

**Comment by alexey-belous** [comment]
I have the same issue but with one file (about 80 MB). It takes about 20 minutes to publish it.

**Comment by MichelZ** [comment]
Is this something that can be improved? I don't like the compress/decompress method. Could it be multithreaded or something to speed it up?

**Comment by Sebs030** [comment]
I am having a similar issue, where but on the unzipping part of the deployment. In the build process of a node.js project the `npm install` process takes 20s, the zip Archive Files process takes 11s and the publish artifact 16s, but in the release pipeline in the deployment to the webapp the unzipping process takes unreasonably long, i.e. almost 20mins, but only after I removed the web.config as described here: [azure-app-service-deploy-problem.html](https://developercommunity.visualstudio.com/content/problem/416229/azure-app-service-deploy-problem.html). I guess it´sbecause of all the node_modules being unpacked

**Comment by Sebs030** [comment]
Hi Piotr,
yes agree, it´s weird, I used a predefined azure template for build process (Node+Gulp), but removed the gulp task, since i don´t use it. Please find attached screenshot of the build log and deploy log. Build Process: 
![image](https://user-images.githubusercontent.com/29453191/50968659-0887f800-14dc-11e9-9618-ad2cc50c584e.png)
Deploy Process: 
![image](https://user-images.githubusercontent.com/29453191/50968791-73393380-14dc-11e9-86e5-2b34fc05da69.png)

By just looking at it i realize that the build process is hosted on Ubuntu but, the release process on Hosted VS2017, maybe thats the reason?

**Comment by odministrator** [comment]
> 
> 
> Hi Piotr,
> yes agree, it´s weird, I used a predefined azure template for build process (Node+Gulp), but removed the gulp task, since i don´t use it. Please find attached screenshot of the build log and deploy log. Build Process:
> ![image](https://user-images.githubusercontent.com/29453191/50968659-0887f800-14dc-11e9-9618-ad2cc50c584e.png)
> Deploy Process:
> ![image](https://user-images.githubusercontent.com/29453191/50968791-73393380-14dc-11e9-86e5-2b34fc05da69.png)
> 
> By just looking at it i realize that the build process is hosted on Ubuntu but, the release process on Hosted VS2017, maybe thats the reason?

Hey,
Looking on tasks list I realize that you are using another build step (not created by myself) "Archive files". deploy step is also something created by other guys. It's hard to say why they are working this way.
Just FYI, my tasks are Powershell based and can work on Linux only if you have Powershell installed there.

**Comment by rollercoasterdigital** [comment]
I've had a lot of problems with this too, with the deployment of an Angular web app sometimes taking up to an hour. I've also tried to install the node_modules as part of the deployment stage by using `npm ci --only=production` as part of a deployment script, but this can take a long time (locally, `npm ci` takes 1m20s to run, on the web app it will take up to 25m) and will regularly fair with file access errors.
I'd welcome anything that would improve deployment performance for Node apps. Currently I'm testing using `WEBSITE_LOCAL_CACHE_OPTION` as `Always` becuase I had seen some suggestion that a lot of latency in the file transfer was due to the caching structure.

**Comment by Moke** [comment]
I experienced this same issue using a Microsoft ubuntu hosted pipeline in azure devops.

I resolved my problem by altering my azure-pipelines.yml with replacing the following:


```
- task: PublishBuildArtifacts@1
  displayName: 'PublishBuildArtifacts'
  inputs:
    pathtoPublish: '$(Build.ArtifactStagingDirectory)'
    artifactName: 'drop'
```


with the new PublishPipelineArtifact@0 (documentation: https://docs.microsoft.com/en-us/azure/devops/pipelines/artifacts/pipeline-artifacts?view=azure-devops&tabs=yaml )

```
- task: PublishPipelineArtifact@0
  displayName: 'Publish pipeline artifact'
  inputs:
    artifactName: 'new'
    targetPath: '$(Build.ArtifactStagingDirectory)'
```

After this adjustment my build time went from ~15 minutes to ~2 minutes.

**Comment by gregpakes** [comment]
@moke

My understanding is that Pipeline Artifacts and Build Artifacts are two very different things.

Pipeline Artifacts are only available to you to download in the Pipeline itself.  Build Artifacts are available to you from a Release Pipleline.

This is not a solution for most people.

**Comment by 3bdNKocY** [comment]
@Moke Your recommendation also works for on-prem builds.
My build times dropped from 14m to sub 2m builds.

**Comment by stofte** [comment]
A project with many files, who could possibly have imagined that in azure devops? I'm shocked there's still no built in fix, instead we must all juggle zip files and whatnot.

For a few of our larger projects, upping/downing the artifact dominates all other tasks, easily being 50 to 100% extra times for simply copying files (which is nothing to some numbers in this thread, thankfully, we just accept Azure is slow as fuck).

**Comment by gregpakes** [comment]
@stofte

1. Presumably you missed the previous comments saying that new Publish Pipeline Artifact task solves this issue.
2. This issue really has nothing to do with Azure being slow.  It’s about the mechanics of the artifact upload. It would be slow on any provider.

**Comment by stofte** [comment]
@gregpakes I don't see those comments, could you highlight?

**Comment by PaulVrugt** [comment]
Having the same issue here. I don't think it should matter how many files are part of artifacts, uploading should be fast. We now need to add a workaround by compressing it first, which shouldn't be necessary

**Comment by gregpakes** [comment]
@PaulVrugt Have you tried the new Pipeline Artifact Publish task as discussed above?

**Comment by PaulVrugt** [comment]
@gregpakes 
Yes I have, and it does speed up the process, but I believe it should not be necessary to do this. This should be handled correctly by azure devops itself. Now we are forced to create a workaround in each pipeline.

**Comment by gregpakes** [comment]
@PaulVrugt

How is it a workaround? This new task replaces the old one.  I think this task does exactly what you’re saying. Azure Devops have fixed the issue by creating a new publish task.

Am I missing something?

**Comment by Moke** [comment]
@PaulVrugt I am sorry you feel computer's should function like magic. Unfortunately they do not.

**Comment by gregpakes** [comment]
I am going to close this issue as I feel the new `Publish Pipeline Artifact` task resolves this issue.

**Summary**

The `Publish Build Artifact` has been replaced with the 'Publish Pipeline Artifact` which does not exhibit this issue.

```
- task: PublishPipelineArtifact@0
  displayName: 'Publish pipeline artifact'
  inputs:
    artifactName: 'new'
    targetPath: '$(Build.ArtifactStagingDirectory)'
```

**Comment by PaulVrugt** [comment]
@gregpakes
@moke

I completely misread the last part of this thread. I thought you were talking about the compress/decompress tasks. If the new publish task fixes this, then I agree that fixes the issue. I'll try it later today.

Sorry for the confusion

**Comment by nbandi413** [comment]
> I experienced this same issue using a Microsoft ubuntu hosted pipeline in azure devops.
> 
> I resolved my problem by altering my azure-pipelines.yml with replacing the following:
> 
> ```
> - task: PublishBuildArtifacts@1
>   displayName: 'PublishBuildArtifacts'
>   inputs:
>     pathtoPublish: '$(Build.ArtifactStagingDirectory)'
>     artifactName: 'drop'
> ```
> 
> with the new PublishPipelineArtifact@0 (documentation: https://docs.microsoft.com/en-us/azure/devops/pipelines/artifacts/pipeline-artifacts?view=azure-devops&tabs=yaml )
> 
> ```
> - task: PublishPipelineArtifact@0
>   displayName: 'Publish pipeline artifact'
>   inputs:
>     artifactName: 'new'
>     targetPath: '$(Build.ArtifactStagingDirectory)'
> ```
> 
> After this adjustment my build time went from ~15 minutes to ~2 minutes.

Saved the day for me, thank you very much. 
Dropped from 15 minutes to 1m 35 sec.

- task: PublishPipelineArtifact@1
  inputs:
    targetPath: '$(build.artifactstagingdirectory)'
    artifact: 'dropLikeitsHot'
    publishLocation: 'pipeline'

**Comment by the-vampiire** [comment]
can anyone confirm that 

```yml
- task: PublishPipelineArtifact@0
  displayName: 'Publish pipeline artifact'
  inputs:
    artifactName: 'new'
    targetPath: '$(Build.ArtifactStagingDirectory)'
```

is equivalent to

```yml
- publish: '$(Build.ArtifactStagingDirectory)'
  artifact: 'new'
```

in the [main yaml reference doc they say to use](https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema?view=azure-devops&tabs=schema%2Cparameter-schema#publish) the `publish` abstraction:

> The publish keyword is a shortcut for the Publish Pipeline Artifact task. The task publishes (uploads) a file or folder as a pipeline artifact that other jobs and pipelines can consume.

but when i use this i get `path does not exist`

**Comment by the-vampiire** [comment]
and as a followup it seems it always zips the output dir (even without the zip directive set to true). why arent we required to set the target path as `$(Build.StagingDirectory)/<project name>.zip`?

**Comment by cawoodm** [comment]
Why is there no `PublishBuildArtifacts@0`? `PublishBuildArtifacts@1` is slow as cuss.

What's the difference between `PublishBuildArtifacts` and `PublishPuildArtifacts`??

**Comment by hanssens** [comment]
Wauw, indeed. Changing from PublishBuildArtifacts to PublishPipelineArtifact significantly decreased my waiting time. 

With a pre-zipped artifact of ±120Mb the difference on this single step alone is amazing:

PublishBuildArtifact: 7-15 min.
PublishPipelineArtifact: < 1-1.5 min.

Executed and roughly measured on several runs, on different days, on both hosted and private build server.

**Comment by msherms2** [comment]
I'm assuming it must have changed but per the docs, PublishPipelineArtifacts is not available for On Prem, so I'm surprised to see people here claiming it sped them up. Bummer that they removed it (or people were incorrect) from On Prem.

**Comment by sercandumansiz** [comment]
I know it can be a bit challenging to keep track of all the comments here, so I wanted to share a quick solution. If you're looking to improve your Azure Pipeline, consider changing your step from **Publish Build Artifact** to **Publish Pipeline Artifact**.

This change can help streamline your workflow.

For more details on how to make this switch, check out the official documentation here: [Migrate from Build Artifacts](https://learn.microsoft.com/en-us/azure/devops/pipelines/artifacts/pipeline-artifacts?view=azure-devops&tabs=yaml#migrate-from-build-artifacts).

Hope this helps!

**Comment by msherms2** [comment]
> I know it can be a bit challenging to keep track of all the comments here, so I wanted to share a quick solution. If you're looking to improve your Azure Pipeline, consider changing your step from **Publish Build Artifact** to **Publish Pipeline Artifact**.
> 
> This change can help streamline your workflow.
> 
> For more details on how to make this switch, check out the official documentation here: [Migrate from Build Artifacts](https://learn.microsoft.com/en-us/azure/devops/pipelines/artifacts/pipeline-artifacts?view=azure-devops&tabs=yaml#migrate-from-build-artifacts).
> 
> Hope this helps!

.......

as i just said that isn't available to OnPrem for a very unknown reason, no, it didn't help me much :) I'd like it very much to be a solution, but they don't seem to want to answer why they only enhance the cloud product.

**Comment by msherms2** [comment]
> > as i just said that isn't available to OnPrem for a very unknown reason, no, it didn't help me much :) I'd like it very much to be a solution, but they don't seem to want to answer why they only enhance the cloud product.
> 
> Please raise a new issue for this. I believe there are solutions to your problem, but this thread is not the appropriate home.
> 
> To clarify my good friend's comment, this issue is for the cloud hosted pipelines and an acceptable solution is available. This issue is closed.
> 
> [@gregpakes](https://github.com/gregpakes) Could you please edit your initial posting to include a summary of the solution i previously provided?

https://github.com/microsoft/azure-pipelines-tasks/issues/19698

They ignored it and marked it "stale" without even looking at it :)

Trust me, it's annoying I have to zip up half my artifacts because the download is slightly not as terrible that way. I'm trying everything I can - using this post, and other posts.
